src/data/nat/pochhammer.lean

```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import tactic.abel
+
+/-!
+# The rising and falling factorial functions
+
+We define and prove some basic relations about
+`rising_factorial x n = x * (x+1) * ... * (x + n - 1)`
+which is also known as the Pochhammer function
+and its cousin `falling_factorial`.
+
+## TODO
+There is lots more in this direction:
+* q-factorials, q-binomials, q-Pochhammer.
+* Defining Bernstein polynomials (e.g. as one way to prove Weierstrass' theorem).
+-/
+
+variables {R : Type*}
+
+section
+variables [semiring R]
+
+/--
+The rising factorial function: `rising_factorial x n = x * (x+1) * ... * (x + n - 1)`.
+
+It is also sometimes called the Pochhammer polynomial, or the upper factorial.
+Notations in the mathematics literature vary extensively.
+-/
+def rising_factorial : R → ℕ → R
+| r 0 := 1
+| r (n+1) := r * rising_factorial (r+1) n
+
+@[simp]
+lemma rising_factorial_zero {r : R} : rising_factorial r 0 = 1 := rfl
+@[simp]
+lemma rising_factorial_one {r : R} : rising_factorial r 1 = r := by simp [rising_factorial]
+
+lemma rising_factorial_eq_mul_left {r : R} {n : ℕ} :
+  rising_factorial r (n + 1) = r * rising_factorial (r+1) n := rfl
+
+lemma rising_factorial_eq_mul_right {r : R} {n : ℕ} :
+  rising_factorial r (n + 1) = rising_factorial r n * (r + n) :=
+begin
+  induction n with n ih generalizing r,
+  { simp, },
+  { rw [rising_factorial, ih, rising_factorial, nat.succ_eq_add_one],
+    push_cast,
+    rw [mul_assoc, add_comm (n : R) 1, add_assoc], }
+end
+
+lemma rising_factorial_mul_rising_factorial {r : R} {n m : ℕ} :
+  rising_factorial r n * rising_factorial (r + n) m = rising_factorial r (n + m) :=
+begin
+  induction m with m ih,
+  { simp, },
+  { rw [rising_factorial_eq_mul_right, ←mul_assoc, ih, nat.add_succ, rising_factorial_eq_mul_right],
+    push_cast,
+    rw [add_assoc], }
+end
+end
+
+section
+variables [ring R]
+
+/--
+The falling factorial function: `falling_factorial x n = x * (x-1) * ... * (x - (n - 1))`.
+-/
+def falling_factorial : R → ℕ → R
+| r 0 := 1
+| r (n+1) := r * falling_factorial (r-1) n
```

Comment:
```suggestion
def falling_factorial (x : R) (n : ℕ) : R :=
rising_factorial (x - (n - 1)) n
```
I'm not sure if this simplifies much, but there's lots of duplication between `rising_factorial` and `falling_factorial`.

Reply:
I'm hesitant, but perhaps unreasonably.

This only gives exactly the same definition if either the ring is commutative, or you go to the extra effort of noticing that `x - a` and `x - b` commute when `a b : nat`. (I think someone may eventually want falling factorials in a noncommutative ring, e.g. you sometimes see them in differential operators.)

Reply:
I don't expect the duplication would decrease a lot, since you'd have to prove almost all results twice anyway, once for `nat` and for a ring. And with `rising_factorial_eq_falling_factorial`, it shouldn't be too hard to transfer any remaining properties if needed.